File: typescript/src/patterns/event-sourcing/Account.ts
import type { AccountEvent } from './Events.js';
import { EventStore } from './EventStore.js';

export class Account {
  private balance = 0;
  constructor(private id: string, private store: EventStore) {}

  private apply(e: AccountEvent) {
    switch (e.type) {
      case 'Deposited': this.balance += e.amount; break;
      case 'Withdrawn': this.balance -= e.amount; break;
    }
  }

  load() { this.store.load(this.id).forEach(e => this.apply(e)); return this; }
  getBalance() { return this.balance; }

  deposit(amount: number) {
    if (amount <= 0) throw new Error('amount > 0');
    const e: AccountEvent = { type: 'Deposited', amount };
    this.store.append(this.id, e);
    this.apply(e);
  }
  withdraw(amount: number) {
    if (amount <= 0) throw new Error('amount > 0');
    if (this.balance < amount) throw new Error('insufficient');
    const e: AccountEvent = { type: 'Withdrawn', amount };
    this.store.append(this.id, e);
    this.apply(e);
  }
}

File: python/src/patterns/event_sourcing/account.py
from .event_store import EventStore
from .events import AccountEvent, Deposited, Withdrawn

class Account:
    def __init__(self, id: str, store: EventStore) -> None:
        self.id = id
        self.store = store
        self.balance = 0.0
    def apply(self, e: AccountEvent) -> None:
        if isinstance(e, Deposited):
            self.balance += e.amount
        elif isinstance(e, Withdrawn):
            self.balance -= e.amount
    def load(self) -> "Account":
        for e in self.store.load(self.id):
            self.apply(e)
        return self
    def deposit(self, amount: float) -> None:
        if amount <= 0: raise ValueError("amount > 0")
        e = Deposited(amount)
        self.store.append(self.id, e)
        self.apply(e)
    def withdraw(self, amount: float) -> None:
        if amount <= 0: raise ValueError("amount > 0")
        if self.balance < amount: raise ValueError("insufficient")
        e = Withdrawn(amount)
        self.store.append(self.id, e)
        self.apply(e)
